mesa/experimental/cell_space/grid.py

```diff
@@ -0,0 +1,179 @@
+from random import Random
+
+from mesa.experimental.cell_space import Cell, DiscreteSpace
+
+
+class Grid(DiscreteSpace):
+    """Base class for all grid and network classes
+
+    Attributes:
+        width (int): width of the grid
+        height (int): height of the grid
+        torus (bool): whether the grid is a torus
+        _try_random (bool): whether to get empty cell be repeatedly trying random cell
+
+    """
+
+    def __init__(
+        self,
+        width: int,
+        height: int,
+        torus: bool = False,
+        capacity: int | None = None,
+        random: Random | None = None,
+        cell_klass: type[Cell] = Cell,
+    ) -> None:
+        super().__init__(capacity=capacity, random=random, cell_klass=cell_klass)
+        self.torus = torus
+        self.width = width
+        self.height = height
+        self._try_random = True
+
+    def select_random_empty_cell(self) -> Cell:
+        # FIXME:: currently just a simple boolean to control behavior
+        # FIXME:: basically if grid is close to 99% full, creating empty list can be faster
+        # FIXME:: note however that the old results don't apply because in this implementation
+        # FIXME:: because empties list needs to be rebuild each time
+        # This method is based on Agents.jl's random_empty() implementation. See
+        # https://github.com/JuliaDynamics/Agents.jl/pull/541. For the discussion, see
+        # https://github.com/projectmesa/mesa/issues/1052 and
+        # https://github.com/projectmesa/mesa/pull/1565. The cutoff value provided
+        # is the break-even comparison with the time taken in the else branching point.
+        if self._try_random:
+            while True:
+                cell = self.all_cells.select_random_cell()
+                if cell.is_empty:
+                    return cell
+        else:
+            return super().select_random_empty_cell()
+
+
+class OrthogonalGrid(Grid):
```

Comment:
Maybe I'm nitpicking too much, but isn't a grid by definition `Orthogonal`? If we look at the cambridge definition:

> a [pattern](https://dictionary.cambridge.org/dictionary/english/pattern) or [structure](https://dictionary.cambridge.org/dictionary/english/structure) made from [horizontal](https://dictionary.cambridge.org/dictionary/english/horizontal) and [vertical](https://dictionary.cambridge.org/dictionary/english/vertical) [lines](https://dictionary.cambridge.org/dictionary/english/line) [crossing](https://dictionary.cambridge.org/dictionary/english/crossing) each other to [form](https://dictionary.cambridge.org/dictionary/english/form) [squares](https://dictionary.cambridge.org/dictionary/english/square).

I would expect the maybe the (super)classes `ContiniousSpace` and `DiscreteSpace`, where `Grid` and `Hexagonal` (or `GridSpace` and `HexogonalSpace`) can be subclasses of `DiscreteSpace`.

Reply:
I am adopting cellular automata terminology here according to which "A grid is called orthogonal if all grid lines intersect at a right angle". So in case of hex tiling, you don't have right angle and no orthogonal grid.

Reply:
First you can choose between `DiscreteSpace` and `ContiniousSpace`. `DiscreteSpace` can than be divided in two kinds: `Network` and what's now called `Grid`. Only grid implies squares, so I don't find that a logical name. What the collection of (`Orthogonal`)`Grid`, `Hexagonal` and `Triangular` is, is a division of space into discrete units in 2D space. We could call those Cells, Polygons, Tiles, or anything else. But the superclass of those shouldn't be called Grid, since that's the square thing (what it currently already is).

Reply:
So, grid does not necessarily imply squares. I first thought so as well, but I dove into some math definitions and turned out to be wrong. In various branches of mathematics, a grid is merely a set of intersecting lines.    
According to [Wolfram](https://mathworld.wolfram.com/Grid.html), "a grid usually refers to two or more infinite sets of evenly-spaced parallel lines at particular angles to each other in a plane, or the intersections of such lines".

Orthogonal Grids are grids where the lines intersect at a right angle. Square grids are a subclass of Orthogonal Grids (but this only makes sense if you are tesselating a continuous space). Isometric, or triangular grids have 3 lines intersecting at 60 degrees. Hex grids are formed by tessellating the space with regular hexagons.

Reply:
So than `SquareGrid` would be best, right?

Reply:
No, orthogonal is correct. To establish whether the grid is square or not, you require a cartesian space upon which you impose an orthogonal grid. Depending on the number of grid cells along the x and y axis can you establish the shape of the grid cell, and only then can you establish whether it is a square grid or not.